clients/src/main/resources/common/message/FetchRequest.json

```diff
@@ -100,7 +102,9 @@
         { "name": "LogStartOffset", "type": "int64", "versions": "5+", "default": "-1", "ignorable": true,
           "about": "The earliest available offset of the follower replica.  The field is only used when the request is sent by the follower."},
         { "name": "PartitionMaxBytes", "type": "int32", "versions": "0+",
-          "about": "The maximum bytes to fetch from this partition.  See KIP-74 for cases where this limit may not be honored." }
+          "about": "The maximum bytes to fetch from this partition.  See KIP-74 for cases where this limit may not be honored." },
+        { "name": "ReplicaDirectoryId", "type": "uuid", "versions": "17+", "taggedVersions": "17+", "tag": 0,
```

Comment:
Pardon me, does this new field obstruct new broker from communicating to old quorum controller? For example, the quorum controller is running with 3.6.2. The broker based on this patch will not be able to generate `FetchRequest` for the controller, because the version=15 can't accept non-default `ReplicaDirectoryId`

Reply:
I saw this ticket, but not a lot of detail: https://issues.apache.org/jira/browse/KAFKA-17018

Reply:
Anyway, I guess that is slightly different. We should always be able to write a request for a lower version.

Reply:
> I guess that is slightly different. We should always be able to write a request for a lower version.

yep, that is different. The case we discuss is "KafkaRaftClient" . IMHO, we should update the request if the version is <17. for example:
```java
            if (version < 17) {
                fetchRequestData.topics()
                    .forEach(t -> t.partitions()
                    .forEach(p -> p.setReplicaDirectoryId(Uuid.ZERO_UUID)));
            }
```

https://github.com/apache/kafka/blob/5596f9e1d561dfef46a83fbd7264e6745ca538cc/clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java#L142

Reply:
I file https://issues.apache.org/jira/browse/KAFKA-17276 for it

Reply:
Thanks @chia7712! If you open a PR for this I will take a look since I would like these tests to be fixed.

Reply:
@chia7712 this is a tagged field so it should show as an unknown tagged field on older client. KRaft only support Fetch with version 12+ and tag fields (flexible version) was added in version 12. Are you not seeing that? We should be exercising this in `KafkaRaftClientTest` and `RaftClientTestContext`. This is how we test different version in the KRaft tests using `RaftClientTestContxt.fetchRpcVersion`:

```java
      private short fetchRpcVersion() {
          if (kip853Rpc) {
              return 17;
          } else {
              return 16;
          }
      }
```

What error are you getting @chia7712 ?

Reply:
I see the issue in the generator code:
```java
              if (_version >= 17) {
                  if (!this.replicaDirectoryId.equals(Uuid.ZERO_UUID)) {
                      _numTaggedFields++;
                  }
              } else {
                  if (!this.replicaDirectoryId.equals(Uuid.ZERO_UUID)) {
                      throw new UnsupportedVersionException("Attempted to write a non-default replicaDirectoryId at version " + _version);
                  }
              }
```
We need to mark the field as ignorable. Same for the `FetchSnapshotRequest`.

Are you going to work on the PR @chia7712 ? If so, I can review it when it is ready.

Reply:
Error is in the title of this ticket: https://issues.apache.org/jira/browse/KAFKA-17250.

How are you writing the RPC in non-test files? It seems like you always set directory ID here? https://github.com/apache/kafka/blob/5596f9e1d561dfef46a83fbd7264e6745ca538cc/raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java#L2686

You cannot guarantee the receiver supports the same version as the sender.

Reply:
> We need to mark the field as ignorable. Same for the FetchSnapshotRequest.

yep, you are right. that is the better approach. I will update the jira 

> Are you going to work on the PR @chia7712 ? If so, I can review it when it is ready.

@m1a2st he will file PR after the e2e pass on our local.